parser_library/src/parsing/parser_impl.cpp

```diff
@@ -14,453 +14,4604 @@
+                break;
+
+            case U'(': {
+                std::vector<semantics::concat_chain> nested;
+                ccb.push_last_text();
+                if (auto [error] = process_macro_list(nested); error)
+                    return failure;
+                ccb.emplace_back(std::in_place_type<semantics::sublist_conc>, std::move(nested));
+                break;
+            }
+
+            case U'\'':
+                if (auto [error] = lex_macro_operand_string(ccb); error)
+                    return failure;
+
+                next_char_special = false;
+                break;
+
+            case U'&': {
+                if (auto [error] = lex_macro_operand_amp(ccb); error)
+                    return failure;
+                if (op_name && follows<U'='>())
+                    ccb.push_equals<true>();
+                else
+                    next_char_special = false;
+                break;
+            }
+
+            case U'O':
+            case U'S':
+            case U'I':
+            case U'L':
+            case U'T':
+            case U'o':
+            case U's':
+            case U'i':
+            case U'l':
+            case U't':
+                if (!last_char_special)
+                {
+                    consume_into(ccb.last_text_value());
+                    next_char_special = false;
+                    break;
+                }
+                if (auto [error, ncs] = lex_macro_operand_attr(ccb); error)
+                    return failure;
+                else
+                    next_char_special = ncs;
+                break;
+
+            default:
+                next_char_special = !is_ord();
+                consume_into(ccb.last_text_value());
+                break;
+        }
+        op_name = false;
+    }
+}
+
+void parser2::process_optional_line_remark()
+{
+    if (follows<U' '>() && before_nl())
+    {
+        lex_line_remark();
+
+        adjust_lines();
+    }
+}
+
+result_t<void> parser2::process_macro_list(std::vector<semantics::concat_chain>& cc)
+{
+    assert(follows<U'('>());
+
+    consume(hl_scopes::operator_symbol);
+    if (try_consume<U')'>(hl_scopes::operator_symbol))
+        return {};
+
+    if (auto [error] = lex_macro_operand(cc.emplace_back(), true, false); error)
+        return failure;
+
+    while (try_consume<U','>(hl_scopes::operator_symbol))
+    {
+        process_optional_line_remark();
+        if (auto [error] = lex_macro_operand(cc.emplace_back(), true, false); error)
+            return failure;
+    }
+
+    if (!match<U')'>(hl_scopes::operator_symbol, diagnostic_op::error_S0011))
+        return failure;
+
+    return {};
+}
+
+result_t<void> parser2::handle_initial_space(bool reparse)
+{
+    if (!reparse && *input.next != U' ')
+    {
+        add_diagnostic(diagnostic_op::error_S0002);
+        consume_rest();
+        return failure;
+    }
+
+    // skip spaces
+    while (follows<U' '>())
+        consume();
```

Comment:
I'm not sure about the impact on readability nor on performance but consider creating a function that would do both the `follows()` check and the `consume()` operation.

The function could also take a lambda operator so we could do other other required operations if needed.

However, I suspect that it might not be worth it to spend time on this as it probably doesn't bring that much more clarity into the code.